### PR TITLE
Overload resolution/picking the right interface

### PR DESCRIPTION
**Pull request by atravita-mods**
Hi Shockah!

This PR does the following:

* Narrows what `AreTypesMatching` considers a matching type. Most importantly, it checks if an interface is **mappable**, otherwise the code can't distinguish between `void F(IProxiedInputA a)` and `void F(IProxiedInputB b)`. (This is cached - if you ask for five different methods that use `IProxiedInputA`, it'll only ever check once).
* Implements some overload resolution. The relevant method is `RankMethods` in `TypeUtilities`, but in effect it first filters for methods with all param names matching. Next, it compares methods by checking each param. A method is considered "better" if all of its param types are assignable to the other methods and there's at least one param that's...more assignable?
    - Something I'm not convinced I like here is the exception it throws - I tried to get it to throw an `AmbiguousMatchException` but Sort repackages that.
    - Incidentally, this fixes a bug with Pintail where it wouldn't necessarily pick the right function if you had overridden methods.

One thing I'm considering changing (but will be a separate PR if I do) is the double for loop in the method resolution - I figure it could be more efficient with a `sort` and a `groupby` so it's not checking every method against every other method. (ie `O(nlogn + mlogm)` instead of `O(mn)`)

Anyways, let me know if anything's unclear! (Apologies for the sometimes deranged commit messages).

### CODE REVIEW

File: Pintail/InterfaceOrDelegateProxyFactory.cs [hunk 1]
```diff
@@ -112,45 +113,68 @@ internal void Prepare(ProxyManager<Context> manager, string typeName)
...
-                }
-                foreach (Type interfaceType in baseType.GetInterfaces())
-                {
-                    foreach (var method in FindInterfaceMethods(interfaceType))
-                    {
-                        yield return method;
-                    }
-                }
-            }
+            // crosscheck this.
```
Comment (Shockah) [inline Pintail/InterfaceOrDelegateProxyFactory.cs:126]: I assume this comment is about IsAssignableTo vs IsAssignableFrom?

File: Pintail/InterfaceOrDelegateProxyFactory.cs [hunk 2]
```diff
@@ -112,45 +113,68 @@ internal void Prepare(ProxyManager<Context> manager, string typeName)
...
+                    {
+                        candidates[targetMethod] = positionConversions;
+                    }
                 }
 
-                this.NoMatchingMethodHandler(proxyBuilder, this.ProxyInfo, targetField, glueField, proxyInfosField, proxyMethod);
-                proxyMethodLoopContinue:;
+                if (candidates.Any())
+                {
+                    if (candidates.Count == 1)
```
Comment (Shockah) [inline Pintail/InterfaceOrDelegateProxyFactory.cs]: The RankMethods method already does this kind of early return, no need to add conditional logic here too.

File: Pintail/InterfaceOrDelegateProxyFactory.cs [hunk 3]
```diff
@@ -112,45 +113,68 @@ internal void Prepare(ProxyManager<Context> manager, string typeName)
...
 
             // proxy methods
             IList<ProxyInfo<Context>> relatedProxyInfos = new List<ProxyInfo<Context>>();
             foreach (MethodInfo proxyMethod in allProxyMethods)
             {
+                var candidates = new Dictionary<MethodInfo, TypeUtilities.PositionConversion?[]>();
                 foreach (MethodInfo targetMethod in allTargetMethods)
                 {
-                    var positionConversions = TypeUtilities.MatchProxyMethod(targetMethod, proxyMethod, this.EnumMappingBehavior);
+                    var positionConversions = TypeUtilities.MatchProxyMethod(targetMethod, proxyMethod, this.EnumMappingBehavior, (new[]{this.ProxyInfo.Target.Type, this.ProxyInfo.Proxy.Type}).ToImmutableHashSet() );
```
Comment (Shockah) [inline Pintail/InterfaceOrDelegateProxyFactory.cs]: ImmutableHashSet.Create

File: Pintail/ProxyManager.cs [hunk 4]
```diff
@@ -171,7 +171,7 @@ private static string GetMd5String(string input)
             methodBuilder.SetParameters(argTypes);
 
             ILGenerator il = methodBuilder.GetILGenerator();
-            il.Emit(OpCodes.Ldstr, $"The {proxyInfo.Proxy.Type.GetShortName()} interface defines method {proxyMethod.Name} which doesn't exist in the API.");
+            il.Emit(OpCodes.Ldstr, $"The {proxyInfo.Proxy.Type.GetShortName()} interface defines method {proxyMethod.Name} which doesn't exist in the API. (It may depend on an interface that was not mappable).");
```
Comment (Shockah) [inline Pintail/ProxyManager.cs:175]: Was this supposed to be the same message as the one in line 149, or does one of them emphasize on purpose?

File: PintailTests/PintailTests.csproj [hunk 5]
```diff
@@ -4,7 +4,7 @@
     <RootNamespace>Nanoray.Pintail.Tests</RootNamespace>
     <TargetFramework>net6.0</TargetFramework>
     <Nullable>enable</Nullable>
-
+    <OutputType>exe</OutputType>
```
Comment (Shockah) [inline PintailTests/PintailTests.csproj]: Huh?

File: PintailTests/Private/FluentProviderApi.cs [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+        }
+
+        public IProxiedInput[]? ArrayReturn() => this.array;
+
+        public IList<IProxiedInput>? ListReturn() => this.array?.ToList();
+    }
+
+    public interface IFluentProviderApi
+    {
+        public int valueState { get; set; }
```
Comment (Shockah) [inline PintailTests/Private/FluentProviderApi.cs]: Interface members are public by default

File: Pintail/TypeUtilities.cs [hunk 7]
```diff
@@ -1,97 +1,161 @@
...
 using System.Reflection;
 
 namespace Nanoray.Pintail
 {
     internal static class TypeUtilities
     {
-        internal enum MethodTypeMatchingPart { ReturnType, Parameter }
-        internal enum MatchingTypesResult { False, IfProxied, True }
-        internal enum PositionConversion { Proxy }
+        static readonly IDictionary<string, List<Type>> cache = new Dictionary<string, List<Type>>();
```
Comment (Shockah) [inline Pintail/TypeUtilities.cs]: I'm not sure I like the cache being static in TypeUtilities. This means multiple proxy managers (if anyone is insane enough to do that) will use the same cache, and that this actually stops being thread-safe (I don't see any locks around this cache).

I think a better solution would be to pass in some "context" object, which would contain the cache.

File: Pintail/TypeUtilities.cs [hunk 8]
```diff
@@ -134,5 +200,176 @@ internal static MatchingTypesResult AreTypesMatching(Type targetType, Type proxy
...
+
+            // Rule out a few common issues with generics first.
+            if (target.IsGenericType != proxy.IsGenericType)
+                return false;
+
+            if (target.IsGenericType && proxy.IsGenericType && target.GenericTypeArguments.Length != proxy.GenericTypeArguments.Length)
+                return false;
+
+            // check the cache.
+            string cachekey = $"{target.AssemblyQualifiedName ?? $"{target.Assembly.GetName().FullName}??{target.Namespace}??{target.Name}"}@@{enumMappingBehavior:D}@@{assignability:D}"; //sometimes AssemblyQualifiedName is null
```
Comment (Shockah) [inline Pintail/TypeUtilities.cs:251]: Were the inner `??` supposed to be parts of the string, or were they supposed to be the null coalescing operator? Currently they're the former.

Comment (atravita-mods) [reply Pintail/TypeUtilities.cs:251]: Part of the string. I wanted a symbol that wasn't actually used in qualified names normally.

File: PintailTests/PintailTests.csproj (hunk 5 above)
Comment (atravita-mods) [reply PintailTests/PintailTests.csproj]: that might have been a leftover from when I tried to get performance benchmarks working, my apologies!

File: Pintail/InterfaceOrDelegateProxyFactory.cs (hunk 1 above)
Comment (atravita-mods) [reply Pintail/InterfaceOrDelegateProxyFactory.cs:126]: Oh, yeah, just a reminder to make sure it was in the right bloody direction

File: Pintail/ProxyManager.cs [hunk 9]
```diff
@@ -243,6 +244,8 @@ public sealed class ProxyManager<Context>: IProxyManager<Context>
         internal readonly ProxyManagerConfiguration<Context> Configuration;
 		private readonly IDictionary<ProxyInfo<Context>, IProxyFactory<Context>> Factories = new Dictionary<ProxyInfo<Context>, IProxyFactory<Context>>();
```
Comment (Shockah) [inline Pintail/ProxyManager.cs]: Indentation (just making a comment so I don't forget to fix it myself)

File: Pintail/ProxyManager.cs [hunk 10]
```diff
@@ -243,6 +244,8 @@ public sealed class ProxyManager<Context>: IProxyManager<Context>
...
 
+        private readonly ConcurrentDictionary<string, List<Type>> interfaceMappabilityCache = new();
```
Comment (Shockah) [inline Pintail/ProxyManager.cs]: PascalCase (note for myself to fix later)

File: Pintail/ReconstructingProxyFactory.cs [hunk 11]
```diff
@@ -11,10 +14,13 @@ internal class ReconstructingProxyFactory<Context>: IProxyFactory<Context>
         private Func<IProxyManager<Context>, object, object> ProxyFactory = null!;
         private Func<IProxyManager<Context>, object, object> UnproxyFactory = null!;
 
-        internal ReconstructingProxyFactory(ProxyInfo<Context> proxyInfo, ProxyManagerEnumMappingBehavior enumMappingBehavior)
+        private readonly ConcurrentDictionary<string, List<Type>> interfaceMappabilityCache;
```
Comment (Shockah) [inline Pintail/ReconstructingProxyFactory.cs]: .

File: PintailTests/Private/ProviderApiWithOverloads.cs [hunk 12]
```diff
@@ -0,0 +1,115 @@
+using System;
+using System.Reflection;
+using System.Reflection.Emit;
+using System.Text;
+
+namespace Nanoray.Pintail.Tests.Provider
+{
+    public interface IProxiedInput
+    {
+        public string teststring { get; set; }
```
Comment (Shockah) [inline PintailTests/Private/ProviderApiWithOverloads.cs]: Unneeded public modifier (note for myself for later)

File: PintailTests/Private/ProviderApiWithOverloads.cs [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+    public interface IProxiedInput2
+    {
+        public string otherteststring { get; set; }
```
Comment (Shockah) [inline PintailTests/Private/ProviderApiWithOverloads.cs]: .

File: PintailTests/ProxyManagerTests.cs [hunk 14]
```diff
@@ -419,22 +421,196 @@ public void TestApiResultEvent()
...
+            Assert.AreEqual(7, otherconsumer.state);
+            // Assert.AreEqual(10, consumerApi.state); // currently fails.
+        }
+
+        [Test]
+        public void TestWithAbstractClass()
+        {
+            var manager = this.CreateProxyManager(new(
+                proxyObjectInterfaceMarking: ProxyObjectInterfaceMarking.MarkerWithProperty
+                ));
```
Comment (Shockah) [inline PintailTests/ProxyManagerTests.cs]: .

File: Pintail/ProxyManager.cs (hunk 4 above)
Comment (atravita-mods) [reply Pintail/ProxyManager.cs:175]: Those were supposed to be the same message, sorry! I think I changed them at two separate times and didn't realize....

**Pull request merged**